### PR TITLE
Make the resource and param input interactively

### PR DESCRIPTION
**Pull request by piyush-garg**
# Changes

This will add the feature of resource and
param input interactively based on the pipeline
spec.

If user has privded resource or param input
by flag then interactive terminal will not start

Also support param as both array and string

Add respective test and docs

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

### CODE REVIEW

File: docs/man/man1/tkn-pipeline-logs.1 [hunk 1]
```diff
@@ -48,20 +48,20 @@ Show pipeline logs
 
 .SH EXAMPLE
 .PP
-# show logs interactively for no inputs
+# interactive mode: shows logs of the selected pipeline run
     tkn pipeline logs \-n namespace
```
Comment (hrishin) [inline docs/man/man1/tkn-pipeline-logs.1]: could you remove `\` from `\-n`?

File: pkg/cmd/pipeline/start.go [hunk 2]
```diff
@@ -39,13 +43,22 @@ const (
 )
 
 type startOptions struct {
+	params             cli.Params
```
Comment (hrishin) [inline pkg/cmd/pipeline/start.go]: better to rename it something else? it's confusing with pipelinerun `param`

File: pkg/cmd/pipeline/start.go [hunk 3]
```diff
@@ -92,44 +109,210 @@ tkn pipeline start foo --param NAME=VALUE --resource source=scaffold-git  -s Ser
...
+	if err != nil {
+		return nil, err
+	}
+	return pres, nil
+}
+
+func getPipelineResourcesByFormat(resources []v1alpha1.PipelineResource) (ret []resourceOptions) {
+	for _, res := range resources {
+		output := ""
+		if string(res.Spec.Type) == "git" {
```
Comment (hrishin) [inline pkg/cmd/pipeline/start.go]: better to use a `switch` statement

File: pkg/cmd/pipeline/start.go [hunk 4]
```diff
@@ -92,44 +109,210 @@ tkn pipeline start foo --param NAME=VALUE --resource source=scaffold-git  -s Ser
...
+		return err
+	}
+
+	return nil
+}
+
+func (opt *startOptions) getInputResources(resources []resourceOptions, pipeline *v1alpha1.Pipeline) error {
+	var ans string
+	for _, res := range pipeline.Spec.Resources {
+		options := getOptionsByFilter(resources, string(res.Type))
```
Comment (hrishin) [inline pkg/cmd/pipeline/start.go]: later on its better to avoid the repeated computations

File: pkg/cmd/pipeline/start.go (hunk 3 above)
Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: ok, cool I will do that way

File: pkg/cmd/pipeline/start.go (hunk 4 above)
Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: you want me to filter the resource by type also upfront and store all different ?

Comment (hrishin) [reply pkg/cmd/pipeline/start.go]: @piyush-garg yes it is okay for now :smile: . Will fix it if we face any issues down the line

File: pkg/cmd/pipeline/start.go (hunk 3 above)
Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: I have addressed this and updated the PR

File: pkg/cmd/pipeline/start.go (hunk 4 above)
Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: I have addressed this and updated the PR

File: pkg/cmd/pipeline/start.go (hunk 2 above)
Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: I have addressed this and updated the PR

**Comment by piyush-garg** [comment]
/test pull-tekton-cli-unit-tests
--

**Comment by pradeepitm12** [comment]
overall looks good to me

File: pkg/cmd/pipeline/start.go [hunk 5]
```diff
@@ -95,49 +117,223 @@ tkn pipeline start foo --param NAME=VALUE --resource source=scaffold-git  -s Ser
...
+				Out: cmd.OutOrStdout(),
+				Err: cmd.OutOrStderr(),
 			}
-			return startPipeline(cmd.OutOrStdout(), opt, p, args[0])
+			return opt.run(args[0])
 		},
 	}
 
-	c.Flags().StringSliceVarP(&res, "resource", "r", []string{}, "pass the resource name and ref")
-	c.Flags().StringSliceVarP(&params, "param", "p", []string{}, "pass the param")
```
Comment (chmouel) [inline pkg/cmd/pipeline/start.go]: If I understand, this PR is removing those two switches? I think that's quite a change (for at least backward compat/documentation),  there is no way to keep them ?

File: pkg/cmd/pipeline/start.go [hunk 6]
```diff
@@ -95,49 +117,223 @@ tkn pipeline start foo --param NAME=VALUE --resource source=scaffold-git  -s Ser
...
+}
+
+func (opt *startOptions) getInput(pname string) error {
+	cs, err := opt.cliparams.Clients()
+	if err != nil {
+		return err
+	}
+	pres, err := getPipelineResources(cs.Tekton, opt.cliparams.Namespace())
+	if err != nil {
+		fmt.Fprintf(opt.stream.Err, "Failed to list pipelineresources from %s namespace \n", opt.cliparams.Namespace())
```
Comment (chmouel) [inline pkg/cmd/pipeline/start.go]: I just learned that today : https://github.com/golang/go/wiki/CodeReviewComments#error-strings 

(no capitalization on error strings)

Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: Addressed

File: pkg/cmd/pipeline/start.go (hunk 5 above)
Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: We made the resource and param input interactively.

Comment (chmouel) [reply pkg/cmd/pipeline/start.go]: right but you can override it with command line options, isnt it ?

Comment (piyush-garg) [reply pkg/cmd/pipeline/start.go]: Addressed and updated the PR

File: pkg/cmd/pipeline/start.go [hunk 7]
```diff
@@ -33,19 +37,30 @@ var (
...
 	Last               bool
 }
 
+type resourceOptionsFilter struct {
+	git         []string
+	image       []string
+	cluster     []string
+	storage     []string
+	pullRequest []string
+}
```
[the remark this answers is not in the document]
Comment (vdemeester) [reply pkg/cmd/pipeline/start.go]: @hrishin not yet :angel:

[the remark this answers is not in the document]
Comment (hrishin) [reply pkg/cmd/pipeline/start.go]: cool @vdemeester :)

**Pull request merged**